store/tikv/mock-tikv/rpc.go

```diff
@@ -0,0 +1,319 @@
+	return &rpcHandler{
+		cluster:   cluster,
+		mvccStore: mvccStore,
+		storeID:   storeID,
+	}
+}
+
+func (h *rpcHandler) handleRequest(req *kvrpcpb.Request) *kvrpcpb.Response {
+	resp := &kvrpcpb.Response{
+		Type: req.Type,
+	}
+	if err := h.checkContext(req.GetContext()); err != nil {
+		resp.RegionError = err
+		return resp
+	}
+	switch req.GetType() {
+	case kvrpcpb.MessageType_CmdGet:
+		resp.CmdGetResp = h.onGet(req.CmdGetReq)
+	case kvrpcpb.MessageType_CmdScan:
+		resp.CmdScanResp = h.onScan(req.CmdScanReq)
+	case kvrpcpb.MessageType_CmdPrewrite:
+		resp.CmdPrewriteResp = h.onPrewrite(req.CmdPrewriteReq)
+	case kvrpcpb.MessageType_CmdCommit:
+		resp.CmdCommitResp = h.onCommit(req.CmdCommitReq)
+	case kvrpcpb.MessageType_CmdCleanup:
+		resp.CmdCleanupResp = h.onCleanup(req.CmdCleanupReq)
+	case kvrpcpb.MessageType_CmdCommitThenGet:
+		resp.CmdCommitGetResp = h.onCommitThenGet(req.CmdCommitGetReq)
+	case kvrpcpb.MessageType_CmdRollbackThenGet:
+		resp.CmdRbGetResp = h.onRollbackThenGet(req.CmdRbGetReq)
+	case kvrpcpb.MessageType_CmdBatchGet:
+		resp.CmdBatchGetResp = h.onBatchGet(req.CmdBatchGetReq)
+	}
+	return resp
+}
+
+func (h *rpcHandler) checkContext(ctx *kvrpcpb.Context) *errorpb.Error {
+	region, leaderID := h.cluster.GetRegion(ctx.GetRegionId())
+	// No region found.
+	if region == nil {
+		return &errorpb.Error{
+			Message: proto.String("region not found"),
+			RegionNotFound: &errorpb.RegionNotFound{
+				RegionId: proto.Uint64(ctx.GetRegionId()),
+			},
+		}
+	}
+	var found bool
+	for _, id := range region.StoreIds {
+		if id == h.storeID {
+			found = true
+			break
+		}
+	}
+	// We got the Region, but not in current Store. Also return a NotFound.
+	if !found {
+		return &errorpb.Error{
+			Message: proto.String("region not found"),
+			RegionNotFound: &errorpb.RegionNotFound{
+				RegionId: proto.Uint64(ctx.GetRegionId()),
+			},
+		}
+	}
+	// No leader.
+	if leaderID == 0 {
+		return &errorpb.Error{
+			Message: proto.String("no leader"),
+			NotLeader: &errorpb.NotLeader{
+				RegionId: proto.Uint64(ctx.GetRegionId()),
+			},
+		}
+	}
+	// Leader is not this Store.
+	if leaderID != h.storeID {
+		return &errorpb.Error{
+			Message: proto.String("not leader"),
+			NotLeader: &errorpb.NotLeader{
+				RegionId:      proto.Uint64(ctx.GetRegionId()),
+				LeaderStoreId: proto.Uint64(leaderID),
+			},
+		}
+	}
+	// Region epoch does not match.
+	if !proto.Equal(region.GetRegionEpoch(), ctx.GetRegionEpoch()) {
+		return &errorpb.Error{
+			Message:    proto.String("stale epoch"),
+			StaleEpoch: &errorpb.StaleEpoch{},
+		}
+	}
+	h.startKey, h.endKey = region.StartKey, region.EndKey
+	return nil
+}
+
+func (h *rpcHandler) keyInRegion(key []byte) bool {
+	return regionContains(h.startKey, h.endKey, key)
+}
+
+func (h *rpcHandler) onGet(req *kvrpcpb.CmdGetRequest) *kvrpcpb.CmdGetResponse {
+	if !h.keyInRegion(req.Key) {
+		panic("onGet: key not in region")
```

Comment:
@shenli If the request passed the Region epoch check, but contains some keys not in the Region, there must be a bug in client's key split logic. I think panic could help us finding bug out as soon as possible.